dev/io.openliberty.microprofile.health.3.0.internal/src/io/openliberty/microprofile/health30/internal/HealthCheck30HttpResponseBuilder.java

```diff
@@ -90,42 +98,40 @@ public void handleUndeterminedResponse(HttpServletResponse httpResponse) {
 
     protected void setChecks(HealthCheckResponse response) {
 
-        HashMap<String, Object> check = new HashMap<String, Object>();
-        check.put(HealthCheckConstants.HEALTH_CHECK_PAYLOAD_NAME, response.getName());
+        JsonObjectBuilder checkBuilder = jsonBuilderFactory.createObjectBuilder();
+
+        Optional<Map<String, Object>> data = response.getData();
+        if ((data != null) && data.isPresent()) {
+            JsonObjectBuilder dataBuilder = jsonBuilderFactory.createObjectBuilder();
+            for (Map.Entry<String, Object> entry : data.get().entrySet()) {
+                dataBuilder.add(entry.getKey(), entry.getValue().toString());
+            }
+            checkBuilder.add(HealthCheckConstants.HEALTH_CHECK_PAYLOAD_DATA, dataBuilder.build());
+        }
+
+        checkBuilder.add(HealthCheckConstants.HEALTH_CHECK_PAYLOAD_NAME, response.getName());
 
         Status checkStatus = response.getStatus();
-        check.put(HealthCheckConstants.HEALTH_CHECK_PAYLOAD_STATUS, checkStatus);
+        checkBuilder.add(HealthCheckConstants.HEALTH_CHECK_PAYLOAD_STATUS, checkStatus.toString());
         if (checkStatus != null) {
             if (checkStatus.equals(Status.DOWN))
                 overallStatus = Status.DOWN;
-        } else {
-            if (tc.isDebugEnabled())
-                Tr.debug(tc, "setChecks(): checkStatus is null");
-            overallStatus = Status.DOWN; // treat as fail case
         }
 
-        Optional<Map<String, Object>> data = response.getData();
-        if ((data != null) && data.isPresent())
-            check.put(HealthCheckConstants.HEALTH_CHECK_PAYLOAD_DATA, data.get());
-
-        checks.add(check);
-        if (tc.isDebugEnabled())
+        checks.add(checkBuilder.build());
+        if (tc.isDebugEnabled()) {
             Tr.debug(tc, "setChecks(): checks = " + checks);
+        }
     }
 
-    protected void setJSONPayload(Map<String, Object> payload, HttpServletResponse httpResponse) {
+    protected void setJSONPayload(JsonObject payload, HttpServletResponse httpResponse) {
         try {
-            httpResponse.getOutputStream().write(json.asBytes(payload));
+            httpResponse.getOutputStream().write(payload.toString().getBytes());
```

Comment:
If all of the logic can move from above to only do createObjectBuilder(Map).build().toString.getBytes() here, that seems ideal and keeps the logic similar to what it was before.  Not sure if that gives you what you need or not.

Of course with older versions where 1.0 doesn't have the createObjectBuilder(Map) method you would need to do things differently.